src/js/common/validate.js

```diff
@@ -0,0 +1,343 @@
+    }
+
+    // The last segment will not have two parts and can be added in its entirety
+    segments.push(gpii.schema.validator.sanitizePathSegment(remainingPath));
+
+    return segments;
+};
+
+/*
+
+  Convenience function to extract the path segments from the `error` data structure returned by AJV.
+
+ */
+gpii.schema.validator.extractPathSegmentsFromError = function (error) {
+    return gpii.schema.validator.extractPathSegments(error.dataPath);
+};
+
+
+/*
+
+  Because dots are allowed in Javascript/JSON object keys, any segment with a dot is escaped by AJV.  The simplest
+  example is `this.that`, which would be escaped as `['this.that']`.  This gives us three additional special characters,
+  namely the square brackets and the single quote.
+
+  If any of those are used, additional escaping is introduced, so `['this.that']` becomes `['[\\'this.that\\']']`.  This
+  introduces the final special character, namely the backslash.
+
+  If an escaped backslash is used, it is double escaped, so that `a\\b` becomes `a\\\\b`.
+
+  This function takes care of reversing this process including all combinations, to ensure that we end up with the
+  original unescaped literal key.
+
+ */
+gpii.schema.validator.sanitizePathSegment = function (segment) {
+    // Discard any leading dot
+    var segmentMinusLeadingDot = segment.replace(/^\./, "");
+
+    // If we are surrounded by `['']`, extract the inner content and then unescape it.
+    var hasSpecialRegexp = /\['(.+)'\]/;
+    var specialMatches = segmentMinusLeadingDot.match(hasSpecialRegexp);
+    if (specialMatches) {
+        // unescape the remaining variables by using them to create a new `String`.
+        return String(specialMatches[1]);
+    }
+
+    return segmentMinusLeadingDot;
+};
+
+// Resolve the underlying data from a hierarchical object using an array of path segments. Returns the portion of the
+// original object at the selected path.  As an example:
+//
+// `resolveOrCreateTargetFromPath({ one: { two: { three: { four: "five"}}}}, ["one", "two", "three"])`
+//
+// Should return:
+//
+// `{ four: "five" }`
+//
+// Note that the relevant portion of the original object is returned, and not just the value.  If
+// `createMissingSegments` is set, the deep structure will be created if it doesn't already exist, Thus:
+//
+// `resolveOrCreateTargetFromPath({}, ["one","two","three"], true)`
+//
+// Will return:
+//
+// `{ one: { two: { three: [] }}}`
+//
+// If you try to resolve a path that does not exist and `createMissingSegments` is not set, `undefined` will be returned.
+//
+gpii.schema.validator.resolveOrCreateTargetFromPath = function (target, path, createMissingSegments) {
+    var resolvedTarget = target;
+    for (var a = 0; a < path.length; a++) {
+        var segment = path[a];
+        if (!resolvedTarget[segment]) {
+            if (createMissingSegments) {
+                resolvedTarget[segment] = a === path.length - 1 ? [] : {};
```

Comment:
If it weren't for this strange rule about how to fill in missing elements, this could just be replaced by fluid.get or fluid.getSimple - why do we have this rule?